In [1]:
import pandas as pd
import numpy as np

In [2]:
nombre = ['Oliva', 'Daniela', 'Juan', 'Germán', 'Edward', 'Alex', 'Julio', 
          'Edgar', 'Angie', 'Irlesa']
puntaje = [11.5, 8, 15.5, np.nan, 8, 19, 13.5, np.nan, 8, 18]
intentos = [1, 3, 2, 3, 2, 3, 1, 1, 2, 1]
califico = ['Sí', 'No', 'Sí', 'No', 'No', 'Sí', 'Sí', 'No', 'No', 'Sí']

indices = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

jugadores = {'nombre': nombre, 'puntaje': puntaje, 'intentos': intentos, 
             'califico': califico}

df = pd.DataFrame(data=jugadores, index=indices)


# Ejercicio 51

Obtener las primeras `n` filas de un objeto `DataFrame`.




In [3]:
df.head(5)

nombre  puntaje  intentos califico
a    Oliva     11.5         1       Sí
b  Daniela      8.0         3       No
c     Juan     15.5         2       Sí
d   Germán      NaN         3       No
e   Edward      8.0         2       No

In [4]:
# `df.iloc[0:5]` está seleccionando las primeras 5 filas del DataFrame `df` utilizando la indexación basada en enteros.
df.iloc[0:5]

nombre  puntaje  intentos califico
a    Oliva     11.5         1       Sí
b  Daniela      8.0         3       No
c     Juan     15.5         2       Sí
d   Germán      NaN         3       No
e   Edward      8.0         2       No

# Ejercicio 52

Obtener las últimas `n` filas de un objeto `DataFrame`.



In [5]:
df.tail(5)

nombre  puntaje  intentos califico
f    Alex     19.0         3       Sí
g   Julio     13.5         1       Sí
h   Edgar      NaN         1       No
i   Angie      8.0         2       No
j  Irlesa     18.0         1       Sí

In [6]:
# `df.iloc[-5:]` está seleccionando las últimas 5 filas del DataFrame `df`.
df.iloc[-5:]

nombre  puntaje  intentos califico
f    Alex     19.0         3       Sí
g   Julio     13.5         1       Sí
h   Edgar      NaN         1       No
i   Angie      8.0         2       No
j  Irlesa     18.0         1       Sí

# Ejercicio 53

Obtener la representación en **NumPy** de un objeto `DataFrame`.


In [7]:
# `df.to_numpy()` es un método en pandas que convierte un objeto DataFrame `df` en una matriz NumPy.
df.to_numpy()

array([['Oliva', 11.5, 1, 'Sí'],
       ['Daniela', 8.0, 3, 'No'],
       ['Juan', 15.5, 2, 'Sí'],
       ['Germán', nan, 3, 'No'],
       ['Edward', 8.0, 2, 'No'],
       ['Alex', 19.0, 3, 'Sí'],
       ['Julio', 13.5, 1, 'Sí'],
       ['Edgar', nan, 1, 'No'],
       ['Angie', 8.0, 2, 'No'],
       ['Irlesa', 18.0, 1, 'Sí']], dtype=object)

# Ejercicio 54

Seleccionar columnas de un objeto `DataFrame` usando notación de slicing.


In [8]:
df.iloc[:,2:5] # desde la columna 3 a la 5

intentos califico
a         1       Sí
b         3       No
c         2       Sí
d         3       No
e         2       No
f         3       Sí
g         1       Sí
h         1       No
i         2       No
j         1       Sí

In [9]:
# `df[['intentos','califico']]` está seleccionando las columnas 'intentos' y 'califico' del DataFrame `df`.
df[['intentos','califico']]

intentos califico
a         1       Sí
b         3       No
c         2       Sí
d         3       No
e         2       No
f         3       Sí
g         1       Sí
h         1       No
i         2       No
j         1       Sí

In [10]:
df.agg({'intentos':'mean','puntaje':('max','idxmax')})


intentos puntaje
mean         1.9     NaN
max          NaN    19.0
idxmax       NaN       f

# Ejercicio 55

Obtener los ejes (atributos de filas y columnas) de un `DataFrame`.


In [11]:
df.axes

[Index(['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j'], dtype='object'),
 Index(['nombre', 'puntaje', 'intentos', 'califico'], dtype='object')]

In [12]:
df.query('nombre.str.contains("a")')
df.query('nombre.str.startswith("C") or  valores> 4')

# Ejercicio 56

Seleccionar filas y columnas específicas de un `DataFrame` con `iloc`.


In [13]:
df.iloc[4,3]

'No'

In [14]:
df.iloc[[1,4],[1,3]] # Fila 2 a fila 3 columna 2 a columna 3

puntaje califico
b      8.0       No
e      8.0       No

# Ejercicio 57

Obtener la dimensión, el tamaño, y la forma de un objeto `DataFrame`.


In [15]:
print(f'Dimension: {df.ndim} Tamaño: {df.size} Forma: {df.shape}')

Dimension: 2 Tamaño: 40 Forma: (10, 4)


In [16]:
df['nombre'].notnull()
df.query('nombre.str.startswith("B")')
df['nombre'].isin(['Bruno','Carlos'])


a    False
b    False
c    False
d    False
e    False
f    False
g    False
h    False
i    False
j    False
Name: nombre, dtype: bool

# Ejercicio 58

Obtener la cantidad de memoria ocupada por cada columna en un `DataFrame`.


In [17]:
df.memory_usage()

Index       380
nombre       80
puntaje      80
intentos     80
califico     80
dtype: int64

# Ejercicio 59

Seleccionar filas y columnas a partir de sus nombres con `ix` y `loc`.


In [18]:
df.loc[['a','b'],['nombre','intentos']] # Fila a y b columna nombre e intentos

nombre  intentos
a    Oliva         1
b  Daniela         3

In [19]:
df.empty

False

# Ejercicio 61

Seleccionar los registros que tienen un campo nulo (nan).


In [20]:
df['Valores'] = np.nan
df['Valores'][5] = 4

C:\Users\bruno\AppData\Local\Temp\ipykernel_20696\3492664333.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Valores'][5] = 4


In [21]:
df[df['Valores'].isnull()] # La fila f no se toma porque es null

nombre  puntaje  intentos califico  Valores
a    Oliva     11.5         1       Sí      NaN
b  Daniela      8.0         3       No      NaN
c     Juan     15.5         2       Sí      NaN
d   Germán      NaN         3       No      NaN
e   Edward      8.0         2       No      NaN
g    Julio     13.5         1       Sí      NaN
h    Edgar      NaN         1       No      NaN
i    Angie      8.0         2       No      NaN
j   Irlesa     18.0         1       Sí      NaN

In [22]:
df[df.isnull().any(axis=1)]

nombre  puntaje  intentos califico  Valores
a    Oliva     11.5         1       Sí      NaN
b  Daniela      8.0         3       No      NaN
c     Juan     15.5         2       Sí      NaN
d   Germán      NaN         3       No      NaN
e   Edward      8.0         2       No      NaN
g    Julio     13.5         1       Sí      NaN
h    Edgar      NaN         1       No      NaN
i    Angie      8.0         2       No      NaN
j   Irlesa     18.0         1       Sí      NaN

<!-- df -->

# Ejercicio 62

Crear un objeto `DataFrame` a partir de objetos `Series`.


In [23]:
serie1 = pd.Series(['Mochila','Carpeta','Regla','Mesa','Celular'])
serie2 = pd.Series(['Negro','Rojo','Blanco','Marron','Azul'])

In [24]:
array_nuevo = np.random.randint(100,100000,(5000,10))
array_nuevo = pd.DataFrame(data = array_nuevo)
array_nuevo.apply(lambda x : x/2)

0        1        2        3        4        5        6        7  \
0     23928.0  36031.5  43318.0  24995.0  32180.5  31864.5  38054.5  23804.5   
1     21360.5  45034.5  24373.0   3068.5   9562.5  46525.0   3503.5  18835.5   
2     25209.5   6833.0  17481.0  39799.5  13949.5  39940.5   5974.5  39023.5   
3     42178.5  10311.5  18927.0   3725.5  40614.0  33580.5   4614.0   1402.5   
4     24993.5   2188.5  11444.0  29298.0  44855.0   7836.0  35478.5   1731.5   
...       ...      ...      ...      ...      ...      ...      ...      ...   
4995  31262.5   1594.5  19301.0  39278.0  20839.5  25747.0  46542.0  28753.5   
4996  33737.5  28636.5  29839.5  10612.5   3095.5  19918.5   9012.5  47957.0   
4997  28950.5  44107.5  20035.5  18306.0   8047.5  14599.5    981.0  45164.5   
4998  33970.0  41981.5  19409.5  31990.0  10828.5  15402.0  45435.5  49268.0   
4999  43997.5  15016.5  21337.0  49429.5   4081.0  28009.5   6960.0  46417.5   

            8        9  
0     21561.5  31075.0  
1     16205.5   3033.5  
2     41870.5  29058.0  
3     34750.0   5389.5  
4     44636.0  21669.0  
...       ...      ...  
4995   5323.5  44826.5  
4996  28303.0   1377.0  
4997  22409.0  13395.0  
4998  38850.0  19445.5  
4999  29151.5  14778.0  

[5000 rows x 10 columns]

In [25]:
precios = pd.Series(np.random.sample(5))
precios

0    0.270436
1    0.538056
2    0.606093
3    0.718010
4    0.455793
dtype: float64

In [26]:

indices = ['a','b','c','d','e']
df_elem =pd.DataFrame( {
    'Elemento':serie1,
    'Color' : serie2,
    'Precio' : precios
})
df_elem.index=indices
df_elem

Elemento   Color    Precio
a  Mochila   Negro  0.270436
b  Carpeta    Rojo  0.538056
c    Regla  Blanco  0.606093
d     Mesa  Marron  0.718010
e  Celular    Azul  0.455793

# Ejercicio 63

Cambiar el tipo de dato de una o más columnas con `astype`.

## Solución:


In [27]:
df_elem['Precio'].apply(lambda x:2*2)


a    4
b    4
c    4
d    4
e    4
Name: Precio, dtype: int64

In [28]:
df_elem['Precio'].astype(float)
df_elem

Elemento   Color    Precio
a  Mochila   Negro  0.270436
b  Carpeta    Rojo  0.538056
c    Regla  Blanco  0.606093
d     Mesa  Marron  0.718010
e  Celular    Azul  0.455793

# Ejercicio 64

Seleccionar registros donde los valores de una columna estén en un rango.


In [29]:
df_elem[(df_elem['Precio'] > 0.3) & (df_elem['Precio'] < 0.9)]

Elemento   Color    Precio
b  Carpeta    Rojo  0.538056
c    Regla  Blanco  0.606093
d     Mesa  Marron  0.718010
e  Celular    Azul  0.455793

In [30]:
# El código `df_elem[df_elem['Precio']. between(0.3,0.9)]` está filtrando el DataFrame `df_elem` en función de los valores de la columna 'Precio'. Selecciona filas donde los valores de 'Precio' están entre 0.3 y 0.9 (ambos inclusive).
df_elem[df_elem['Precio'].between(0.3,0.9)]

Elemento   Color    Precio
b  Carpeta    Rojo  0.538056
c    Regla  Blanco  0.606093
d     Mesa  Marron  0.718010
e  Celular    Azul  0.455793

In [31]:
df_elem.query('Precio > 0.3 and Precio < 0.9')

Elemento   Color    Precio
b  Carpeta    Rojo  0.538056
c    Regla  Blanco  0.606093
d     Mesa  Marron  0.718010
e  Celular    Azul  0.455793

# Ejercicio 65

Determinar los datos en un `DataFrame` que son nulos (NaN).

## Solución:

In [32]:
df.isna()

nombre  puntaje  intentos  califico  Valores
a   False    False     False     False     True
b   False    False     False     False     True
c   False    False     False     False     True
d   False     True     False     False     True
e   False    False     False     False     True
f   False    False     False     False    False
g   False    False     False     False     True
h   False     True     False     False     True
i   False    False     False     False     True
j   False    False     False     False     True

# Ejercicio 66

Encontrar los datos en un `DataFrame` que **no** son nulos (NaN).

## Solución:

In [33]:
df.notna()

nombre  puntaje  intentos  califico  Valores
a    True     True      True      True    False
b    True     True      True      True    False
c    True     True      True      True    False
d    True    False      True      True    False
e    True     True      True      True    False
f    True     True      True      True     True
g    True     True      True      True    False
h    True    False      True      True    False
i    True     True      True      True    False
j    True     True      True      True    False

# Ejercicio 67

Acceder un valor o entrada específico de un objeto `DataFrame` usando `at`.

## Solución:

In [34]:
df.at['a','nombre'] # Accedo a una celda especifica

'Oliva'

# Ejercicio 68

Modificar un valor o entrada específico de un objeto `DataFrame` usando `at`.

## Solución:

In [35]:
df.at['a','nombre'] = 'Marcelo'
df

nombre  puntaje  intentos califico  Valores
a  Marcelo     11.5         1       Sí      NaN
b  Daniela      8.0         3       No      NaN
c     Juan     15.5         2       Sí      NaN
d   Germán      NaN         3       No      NaN
e   Edward      8.0         2       No      NaN
f     Alex     19.0         3       Sí      4.0
g    Julio     13.5         1       Sí      NaN
h    Edgar      NaN         1       No      NaN
i    Angie      8.0         2       No      NaN
j   Irlesa     18.0         1       Sí      NaN

In [36]:
# El código agrupa el DataFrame `df` por la columna 'califico' y luego calcula los valores medio, máximo y mínimo de la columna 'puntaje' para cada grupo.
print(df.groupby('califico').agg({'puntaje':('mean','max','min')}))
print(df.groupby('califico')['puntaje'].agg(['mean','max','min']))

         puntaje            
            mean   max   min
califico                    
No           8.0   8.0   8.0
Sí          15.5  19.0  11.5
          mean   max   min
califico                  
No         8.0   8.0   8.0
Sí        15.5  19.0  11.5


In [37]:
df['puntaje'].where(df['califico'] != 'No',5,inplace=True)
df

nombre  puntaje  intentos califico  Valores
a  Marcelo     11.5         1       Sí      NaN
b  Daniela      5.0         3       No      NaN
c     Juan     15.5         2       Sí      NaN
d   Germán      5.0         3       No      NaN
e   Edward      5.0         2       No      NaN
f     Alex     19.0         3       Sí      4.0
g    Julio     13.5         1       Sí      NaN
h    Edgar      5.0         1       No      NaN
i    Angie      5.0         2       No      NaN
j   Irlesa     18.0         1       Sí      NaN

# Ejercicio 69

Modificar un valor o entrada específico de un objeto `DataFrame` usando `loc`.

## Solución:

In [38]:
df.loc['a','nombre'] = 'Carlos'

# Ejercicio 70

Acceder a un valor o entrada específico de un objeto `DataFrame` usando `loc`.

## Solución:

In [39]:
df.loc['a','nombre']

'Carlos'

# Ejercicio 71

Sumar los valores de una columna de un `DataFrame` con `sum`.

## Solución:

In [40]:
df['puntaje'].sum()

102.5

# Ejercicio 72

Acceder a una entrada de un `DataFrame` utilizando índices numéricos con `iat`.

## Solución

In [41]:
df.iat[0,3] # Accedo mediante las posiciones

'Sí'

# Ejercicio 73

Calcular el promedio de los valores de una columna de un `DataFrame`.

## Solución:

In [42]:
df['puntaje'].mean()

10.25

# Ejercicio 74

Insertar una nueva columna en una posición arbitraria con la función insert.

## Solución:

In [43]:
serie1 = np.where(df['califico'] == 'Sí',True,False)
df.insert(2,'Cal',serie1)


In [44]:
df

nombre  puntaje    Cal  intentos califico  Valores
a   Carlos     11.5   True         1       Sí      NaN
b  Daniela      5.0  False         3       No      NaN
c     Juan     15.5   True         2       Sí      NaN
d   Germán      5.0  False         3       No      NaN
e   Edward      5.0  False         2       No      NaN
f     Alex     19.0   True         3       Sí      4.0
g    Julio     13.5   True         1       Sí      NaN
h    Edgar      5.0  False         1       No      NaN
i    Angie      5.0  False         2       No      NaN
j   Irlesa     18.0   True         1       Sí      NaN

# Ejercicio 75

 Agregar una nueva fila a un objeto `DataFrame` por medio de `loc`.

## Solución:

In [45]:
df.loc['k'] = ['Bruno',20.0,True,1,'Sí',20]

In [46]:
df

nombre  puntaje    Cal  intentos califico  Valores
a   Carlos     11.5   True         1       Sí      NaN
b  Daniela      5.0  False         3       No      NaN
c     Juan     15.5   True         2       Sí      NaN
d   Germán      5.0  False         3       No      NaN
e   Edward      5.0  False         2       No      NaN
f     Alex     19.0   True         3       Sí      4.0
g    Julio     13.5   True         1       Sí      NaN
h    Edgar      5.0  False         1       No      NaN
i    Angie      5.0  False         2       No      NaN
j   Irlesa     18.0   True         1       Sí      NaN
k    Bruno     20.0   True         1       Sí     20.0

# Ejercicio 76:

Recorrer cada una de las columnas de un objeto `DataFrame` con items.

## Solución:

In [47]:
for it in df.items(): # ITEMS ME GENERA UN ITERADOR PARA LAS COLUMAS
    print(it[1][2])  # El primer indice seara al encabezado "0" del  resto [1] y el segundo a cada elemento "Fila"

Juan
15.5
True
2
Sí
nan


In [48]:
df[[col for col in df.columns if col == 'nombre']]

nombre
a   Carlos
b  Daniela
c     Juan
d   Germán
e   Edward
f     Alex
g    Julio
h    Edgar
i    Angie
j   Irlesa
k    Bruno

# Ejercicio 77

Eliminar una fila de un `DataFrame` con la función `drop`.

## Solución:

In [49]:
df.drop(index='g') # PARA REFLEJAR EL CAMBIO DEBO USAR INPLACE

nombre  puntaje    Cal  intentos califico  Valores
a   Carlos     11.5   True         1       Sí      NaN
b  Daniela      5.0  False         3       No      NaN
c     Juan     15.5   True         2       Sí      NaN
d   Germán      5.0  False         3       No      NaN
e   Edward      5.0  False         2       No      NaN
f     Alex     19.0   True         3       Sí      4.0
h    Edgar      5.0  False         1       No      NaN
i    Angie      5.0  False         2       No      NaN
j   Irlesa     18.0   True         1       Sí      NaN
k    Bruno     20.0   True         1       Sí     20.0

In [50]:
df.sort_values(['nombre','puntaje'], ascending = False)

nombre  puntaje    Cal  intentos califico  Valores
g    Julio     13.5   True         1       Sí      NaN
c     Juan     15.5   True         2       Sí      NaN
j   Irlesa     18.0   True         1       Sí      NaN
d   Germán      5.0  False         3       No      NaN
e   Edward      5.0  False         2       No      NaN
h    Edgar      5.0  False         1       No      NaN
b  Daniela      5.0  False         3       No      NaN
a   Carlos     11.5   True         1       Sí      NaN
k    Bruno     20.0   True         1       Sí     20.0
i    Angie      5.0  False         2       No      NaN
f     Alex     19.0   True         3       Sí      4.0

# Ejercicio 78

Ordenar en modo descendente un `DataFrame` por diferentes columnas.

## Solución:

In [51]:
df.loc['g','puntaje'] = 17.0
df.loc['g','nombre'] = 'Juan'

In [52]:
df.sort_values('nombre', ascending = False)

nombre  puntaje    Cal  intentos califico  Valores
c     Juan     15.5   True         2       Sí      NaN
g     Juan     17.0   True         1       Sí      NaN
j   Irlesa     18.0   True         1       Sí      NaN
d   Germán      5.0  False         3       No      NaN
e   Edward      5.0  False         2       No      NaN
h    Edgar      5.0  False         1       No      NaN
b  Daniela      5.0  False         3       No      NaN
a   Carlos     11.5   True         1       Sí      NaN
k    Bruno     20.0   True         1       Sí     20.0
i    Angie      5.0  False         2       No      NaN
f     Alex     19.0   True         3       Sí      4.0

In [53]:
df.sort_values(['nombre','puntaje'], ascending = [True,False]) # Al tambien ordenar por puntaje se invirtio el orden que antes

nombre  puntaje    Cal  intentos califico  Valores
f     Alex     19.0   True         3       Sí      4.0
i    Angie      5.0  False         2       No      NaN
k    Bruno     20.0   True         1       Sí     20.0
a   Carlos     11.5   True         1       Sí      NaN
b  Daniela      5.0  False         3       No      NaN
h    Edgar      5.0  False         1       No      NaN
e   Edward      5.0  False         2       No      NaN
d   Germán      5.0  False         3       No      NaN
j   Irlesa     18.0   True         1       Sí      NaN
g     Juan     17.0   True         1       Sí      NaN
c     Juan     15.5   True         2       Sí      NaN

# Ejercicio 79

Iterar con la función `itertuples` un `DataFrame` por filas a modo de tuplas nombradas.

## Solución:

In [54]:
for tup in df.itertuples(): # GENERO UN ITERADOR PARA LAS FILAS
    print(tup.nombre,tup.puntaje)

Carlos 11.5
Daniela 5.0
Juan 15.5
Germán 5.0
Edward 5.0
Alex 19.0
Juan 17.0
Edgar 5.0
Angie 5.0
Irlesa 18.0
Bruno 20.0


# Ejercicio 80

Reemplazar valores de una columna con la función `map` de un `DataFrame`.

## Solución:

In [57]:
df_filter = df.filter(like="nombre")
df_filter

nombre
a   Carlos
b  Daniela
c     Juan
d   Germán
e   Edward
f     Alex
g     Juan
h    Edgar
i    Angie
j   Irlesa
k    Bruno

TypeError: DataFrame.query() missing 1 required positional argument: 'expr'

In [ ]:
df['puntaje'] = df['puntaje'].map(lambda x:x**2)

# Ejercicio 81

Extraer y remover una columna por su nombre a través de la función `pop`.

## Solución:

In [ ]:
califico = df.pop('califico')


In [ ]:
califico

a    Sí
b    No
c    Sí
d    No
e    No
f    Sí
g    Sí
h    No
i    No
j    Sí
k    Sí
Name: califico, dtype: object

# Ejercicio 82

Reemplazar valores de una columna por medio de la función `replace`.

## Solución:

In [63]:
df['Cal'].replace([True,False],['Sí','No'],inplace=True)
df

nombre  puntaje Cal  intentos califico  Valores
a   Carlos     11.5  Sí         1       Sí      NaN
b  Daniela      5.0  No         3       No      NaN
c     Juan     15.5  Sí         2       Sí      NaN
d   Germán      5.0  No         3       No      NaN
e   Edward      5.0  No         2       No      NaN
f     Alex     19.0  Sí         3       Sí      4.0
g     Juan     17.0  Sí         1       Sí      NaN
h    Edgar      5.0  No         1       No      NaN
i    Angie      5.0  No         2       No      NaN
j   Irlesa     18.0  Sí         1       Sí      NaN
k    Bruno     20.0  Sí         1       Sí     20.0

# Ejercicio 83

Mostrar los n primeros y últimos registros de un objeto `DataFrame` con `head` y `tail`.

## Solución:

In [ ]:
df.head(4)

nombre  puntaje Cal  intentos  Valores
a   Carlos   132.25  Sí         1      NaN
b  Daniela    25.00  No         3      NaN
c     Juan   240.25  Sí         2      NaN
d   Germán    25.00  No         3      NaN

In [ ]:
df.tail(4)

nombre  puntaje Cal  intentos  Valores
h   Edgar     25.0  No         1      NaN
i   Angie     25.0  No         2      NaN
j  Irlesa    324.0  Sí         1      NaN
k   Bruno    400.0  Sí         1     20.0

# Ejercicio 84

Validar con `isin` si un elemento se halla dentro de un `DataFrame`.

## Solución:

In [61]:
df.isin(['Bruno','No']) # Le puedo pasar una serie y me dice true o false en el df si encuentra los valores

nombre  puntaje    Cal  intentos  califico  Valores
a   False    False  False     False     False    False
b   False    False  False     False      True    False
c   False    False  False     False     False    False
d   False    False  False     False      True    False
e   False    False  False     False      True    False
f   False    False  False     False     False    False
g   False    False  False     False     False    False
h   False    False  False     False      True    False
i   False    False  False     False      True    False
j   False    False  False     False     False    False
k    True    False  False     False     False    False

In [ ]:
df['nombre'].isin(['Carlos','Daniela','Juan']) # PUEdo pasarle una serie por parametro y preguntar si es igual

a     True
b     True
c     True
d    False
e    False
f    False
g     True
h    False
i    False
j    False
k    False
Name: nombre, dtype: bool

In [ ]:
df[df['nombre'].isin(['Carlos','Daniela','Juan'])]

nombre  puntaje Cal  intentos  Valores
a   Carlos   132.25  Sí         1      NaN
b  Daniela    25.00  No         3      NaN
c     Juan   240.25  Sí         2      NaN
g     Juan   289.00  Sí         1      NaN

# Ejercicio 85

Reemplazar valores que no cumplan una condición a través de la función `DataFrame.where`.

## Solución:

In [ ]:
df['puntaje'].where(df['puntaje']>200,0,inplace = True) # LOS QUE NO CUMPLAN LA CONDICION SE MODIFICAN
# EL METODO WHERE TAMBIEN PUEDE SER APLICADO A TODO EL DF  CAMBIANDO LOS VALORES DE TODAS LAS COLUMNAS CUANDO EL VALOR EN LA COLUMNA INDICADA CUMPLA LA CONDICION
# EJ df.where(df['puntaje']>200,0,inplace = True) PONE 0 EN TODAS LAS FILAS DONDE EL VALOR DE PUNTAJE SEA MENOR A 200

In [ ]:
df # COMO SE VE LA FUNCION ANTERIOR NO MODIFICA
# USAR INPLACE

nombre  puntaje Cal  intentos  Valores
a   Carlos     0.00  Sí         1      NaN
b  Daniela     0.00  No         3      NaN
c     Juan   240.25  Sí         2      NaN
d   Germán     0.00  No         3      NaN
e   Edward     0.00  No         2      NaN
f     Alex   361.00  Sí         3      4.0
g     Juan   289.00  Sí         1      NaN
h    Edgar     0.00  No         1      NaN
i    Angie     0.00  No         2      NaN
j   Irlesa   324.00  Sí         1      NaN
k    Bruno   400.00  Sí         1     20.0

# Ejercicio 87

Usar la función `mask` para modificar los valores que cumplan una condición.

## Solución:

In [ ]:
df.mask(df['puntaje']>200,0) # A LOS QUE CUMPLAN LA CONDICION MODIFICA TODA LA FILA,TAMBIEN USA INPLACE

nombre  puntaje Cal  intentos  Valores
a   Carlos      0.0  Sí         1      NaN
b  Daniela      0.0  No         3      NaN
c        0      0.0   0         0      0.0
d   Germán      0.0  No         3      NaN
e   Edward      0.0  No         2      NaN
f        0      0.0   0         0      0.0
g        0      0.0   0         0      0.0
h    Edgar      0.0  No         1      NaN
i    Angie      0.0  No         2      NaN
j        0      0.0   0         0      0.0
k        0      0.0   0         0      0.0

# Ejercicio 88

 Iterar cada una de las filas de un `DataFrame` con la función `iterrows`.
 
 ## Solución:

In [ ]:
for fila in df.iterrows():
    print(fila)
    print("-"*30)

('a', nombre      Carlos
puntaje        0.0
Cal             Sí
intentos         1
Valores        NaN
Name: a, dtype: object)
------------------------------
('b', nombre      Daniela
puntaje         0.0
Cal              No
intentos          3
Valores         NaN
Name: b, dtype: object)
------------------------------
('c', nombre        Juan
puntaje     240.25
Cal             Sí
intentos         2
Valores        NaN
Name: c, dtype: object)
------------------------------
('d', nombre      Germán
puntaje        0.0
Cal             No
intentos         3
Valores        NaN
Name: d, dtype: object)
------------------------------
('e', nombre      Edward
puntaje        0.0
Cal             No
intentos         2
Valores        NaN
Name: e, dtype: object)
------------------------------
('f', nombre       Alex
puntaje     361.0
Cal            Sí
intentos        3
Valores       4.0
Name: f, dtype: object)
------------------------------
('g', nombre       Juan
puntaje     289.0
Cal            Sí
inte

# Ejercicio 89

Realizar consultas sobre los elementos de un objeto `DataFrame` con la función `query`.

## Solución:

In [70]:
print(df.query('puntaje>200 and Cal.str.contains("N")'))
print("="*50)
print(df.query('intentos == 1'))
print("="*50)
print(df.query('Cal == "Sí"'))
# TAMBIEN QUERY TIENE EL PARAMETRO INPLACE


Empty DataFrame
Columns: [nombre, puntaje, Cal, intentos, califico, Valores]
Index: []
   nombre  puntaje Cal  intentos califico  Valores
a  Carlos     11.5  Sí         1       Sí      NaN
g    Juan     17.0  Sí         1       Sí      NaN
h   Edgar      5.0  No         1       No      NaN
j  Irlesa     18.0  Sí         1       Sí      NaN
k   Bruno     20.0  Sí         1       Sí     20.0
   nombre  puntaje Cal  intentos califico  Valores
a  Carlos     11.5  Sí         1       Sí      NaN
c    Juan     15.5  Sí         2       Sí      NaN
f    Alex     19.0  Sí         3       Sí      4.0
g    Juan     17.0  Sí         1       Sí      NaN
j  Irlesa     18.0  Sí         1       Sí      NaN
k   Bruno     20.0  Sí         1       Sí     20.0


# Ejercicio 90

Realizar operaciones aritméticas básicas entre dos objetos `DataFrame`.

## Solución:

In [ ]:
df1 = pd.DataFrame(dict(x=[1,5,4,7,10],y=[0,23,100,15,9],z=[9,6,76,12,3]))
df2 = pd.DataFrame(dict(u=[91,7,4,2,55],w=[66,1,8,44,2],v=[15,70,14,32,15]))
df1.mul(df1)


x      y     z
0    1      0    81
1   25    529    36
2   16  10000  5776
3   49    225   144
4  100     81     9

# Ejercicio 91

Obtener como una lista Python los nombres de las columnas de un `DataFrame`.

## Solución:

In [ ]:
list(df.columns)

['nombre', 'puntaje', 'Cal', 'intentos', 'Valores']

# Ejercicio 92

Calcular el módulo sobre cada elemento de un objeto `DataFrame`.

## Solución:

In [ ]:
df1.mod(5)

x  y  z
0  1  0  4
1  0  3  1
2  4  0  1
3  2  0  2
4  0  4  3

# Ejercicio 93

Renombrar las columnas de un objeto `DataFrame`.

## Solución:

In [ ]:
df.rename(columns={'Cal':'califico','Valores':'valores'},inplace=True)
df

nombre  puntaje califico  intentos  valores
a   Carlos     0.00       Sí         1      NaN
b  Daniela     0.00       No         3      NaN
c     Juan   240.25       Sí         2      NaN
d   Germán     0.00       No         3      NaN
e   Edward     0.00       No         2      NaN
f     Alex   361.00       Sí         3      4.0
g     Juan   289.00       Sí         1      NaN
h    Edgar     0.00       No         1      NaN
i    Angie     0.00       No         2      NaN
j   Irlesa   324.00       Sí         1      NaN
k    Bruno   400.00       Sí         1     20.0

# Ejercicio 94

Calcular el producto punto entre dos objetos `DataFrame` con la función `dot`.

## Solución:

In [ ]:
df1['x'].dot(df2['u'])

706

In [77]:
arr1 = np.random.randint(1,500,(20,30))
arr2 = np.random.randint(1,50,(30,10))
print(np.matmul(arr1,arr2))


[[157353 160235 163525 160842 173174 189417 165859 220232 188367 167814]
 [154423 162980 168398 185504 150274 191648 177537 208536 198311 183075]
 [126995 124804 154796 134810 121995 134651 161846 173384 156041 155743]
 [186152 182155 170848 177027 171457 196238 212516 237527 217522 203315]
 [215885 189443 210404 197775 199993 224742 227769 275726 243462 207692]
 [150563 162413 160813 171090 162476 191024 165707 215577 185406 161138]
 [199296 182707 187722 172754 175858 193816 206541 242161 223479 190638]
 [173728 166057 178445 159346 165109 212809 192599 205221 199170 178278]
 [129280 144612 146592 146060 136081 161507 178168 176395 182921 156484]
 [190438 171027 180616 150951 175749 184249 183924 241194 203203 194585]
 [130718 153972 151767 154058 135158 155898 181299 173479 193334 158148]
 [177520 184402 181614 175125 169366 209183 203025 227309 202863 184807]
 [160493 187806 176477 173663 162904 179858 162683 225038 205767 184051]
 [163623 178678 177863 179773 173831 209068 189936 

# Ejercicio 95

Seleccionar registros donde se cumpla una condición para valores de una columna.

## Solución:

In [79]:
df.loc[df['puntaje']>200]

Empty DataFrame
Columns: [nombre, puntaje, Cal, intentos, califico, Valores]
Index: []

# Ejercicio 96

Usar operadores relacionales para comparar elementos de objetos `DataFrame`.

## Solución:

In [ ]:
df1.isin(df2)

x      y      z
0  False  False  False
1  False  False  False
2  False  False  False
3  False  False  False
4  False  False  False

# Ejercicio 97

Usar métodos de instancia de `DataFrame` para comparacionese de igualdad.

## Solución:

In [ ]:
df1.eq(df2) # equal

u      v      w      x      y      z
0  False  False  False  False  False  False
1  False  False  False  False  False  False
2  False  False  False  False  False  False
3  False  False  False  False  False  False
4  False  False  False  False  False  False

In [ ]:
df1.ne(df2) # not equal

u     v     w     x     y     z
0  True  True  True  True  True  True
1  True  True  True  True  True  True
2  True  True  True  True  True  True
3  True  True  True  True  True  True
4  True  True  True  True  True  True

In [ ]:
df1.le(df2) #less equal,ge

u      v      w      x      y      z
0  False  False  False  False  False  False
1  False  False  False  False  False  False
2  False  False  False  False  False  False
3  False  False  False  False  False  False
4  False  False  False  False  False  False

In [ ]:
df1.lt(df2) # less then,gt

u      v      w      x      y      z
0  False  False  False  False  False  False
1  False  False  False  False  False  False
2  False  False  False  False  False  False
3  False  False  False  False  False  False
4  False  False  False  False  False  False

# Ejercicio 98

Cambiar el orden de dos columnas de un objeto `DataFrame`.

## Solución:

In [ ]:
df.sort_index(axis=1)

califico  intentos   nombre  puntaje  valores
a       Sí         1   Carlos     0.00      NaN
b       No         3  Daniela     0.00      NaN
c       Sí         2     Juan   240.25      NaN
d       No         3   Germán     0.00      NaN
e       No         2   Edward     0.00      NaN
f       Sí         3     Alex   361.00      4.0
g       Sí         1     Juan   289.00      NaN
h       No         1    Edgar     0.00      NaN
i       No         2    Angie     0.00      NaN
j       Sí         1   Irlesa   324.00      NaN
k       Sí         1    Bruno   400.00     20.0

# Ejercicio 101

Aplicar una función a todos los elementos de un `DataFrame` con la función `apply`.

## Solución:

In [ ]:
df1.apply(lambda x:x**2)

x      y     z
0    1      0    81
1   25    529    36
2   16  10000  5776
3   49    225   144
4  100     81     9

# Ejercicio 102

Guardar los registros de un `DataFrame` en un archivo de formato **CSV**.

## Solución:

In [ ]:
df.to_csv('datos.csv')